torch/utils/data/dataset.py

```diff
@@ -206,6 +208,50 @@ def __len__(self):
         return self.tensors[0].size(0)
 
 
+class StackDataset(Dataset[Union[Tuple[T_co, ...], Dict[str, T_co]]]):
+    r"""Dataset as a stacking of multiple datasets.
+
+    This class is useful to assemble different parts of complex input data, given as datasets.
+
+    Example:
+        >>> # xdoctest: +SKIP
+        >>> images = ImageDataset()
+        >>> texts = TextDataset()
+        >>> tuple_stack = StackDataset(images, texts)
+        >>> tuple_stack[0] == (images[0], texts[0])
+        >>> dict_stack = StackDataset(image=images, text=texts)
+        >>> dict_stack[0] == {'image': images[0], 'text': texts[0]}
+
+    Args:
+        *args (Dataset): Datasets for stacking returned as tuple.
+        **kwargs (Dataset): Datasets for stacking returned as dict.
+    """
+    datasets: Union[Tuple[Dataset[T_co], ...], Dict[str, Dataset[T_co]]]
+
+    def __init__(self, *args: Dataset[T_co], **kwargs: Dataset[T_co]) -> None:
+        if args:
+            assert not kwargs, "Supported only tuple or dict like output setup, but both given"
+            assert all(len(args[0]) == len(dataset) for dataset in args), "Size mismatch between datasets"  # type: ignore[arg-type]
+
+            self.datasets = args
+        else:
+            assert kwargs, "At least one dataset should be passed"
+            tmp = list(kwargs.values())
+            assert all(len(tmp[0]) == len(dataset) for dataset in tmp), "Size mismatch between datasets"  # type: ignore[arg-type]
+
+            self.datasets = kwargs
+
+    def __getitem__(self, index):
+        if isinstance(self.datasets, dict):
+            return {k: dataset[index] for k, dataset in self.datasets.items()}
```

Comment:
Aha, I am sorry that I should clarify this part. Should we expect it to return Tuple if `*args` is provided or Dict if `**kwargs` is provided?

Reply:
if args filled the output will be a tuple
if kwargs filled -> dict

Reply:
Sounds great!